Lab_08/src/Main.java

```diff
@@ -5,212 +5,212 @@
-                for (String s : list) {
-                    all.append(s);
-                }
-                all.append(" | ");
-            }
-            return "\t" + key + " -> " + all.substring(0, all.length() - 2);
-        } else return "NonTerminal not found\n";
-    }
+               System.out.println("Sequence " + w + " is not accepted.");
+          }
+     }
+
+     private static String displayProductions(Stack<String> pi) {
+          StringBuilder sb = new StringBuilder();
+
+          for (String productionIndexString : pi) {
+               if (productionIndexString.equals("~")) {
+                    continue;
+               }
+               Integer productionIndex = Integer.parseInt(productionIndexString);
+               ll1.getProductionsNumbered().forEach((key, value) -> {
+                    if (productionIndex.equals(value))
+                         sb.append(value).append(": ").append(key.getKey()).append(" -> ").append(key.getValue()).append("\n");
+               });
+          }
+
+          return sb.toString();
+     }
+
+     private static void printMenu() {
+          System.out.println("------------ MENU ------------");
+          String menu = "";
+          menu += "1. Display non-terminals\n";
+          menu += "2. Display terminals\n";
+          menu += "3. Display all productions\n";
+          menu += "4. Display all productions for a non-terminal\n";
+          menu += "5. Display parse table\n";
+          menu += "6. Parse sequence\n";
+          menu += "7. Parse seq.in for g1.in\n";
+          menu += "8. Parse pif.out for g2.in\n";
+          menu += "0. Exit";
+          System.out.println(menu);
+          System.out.println("------------------------------");
+     }
+
+     private static void printNonTerminals() {
+          System.out.println("N: " + grammar.getN());
+     }
+
+     private static void printTerminals() {
+          System.out.println("E: " + grammar.getE());
+     }
+
+     private static void printProductions() {
+          System.out.println("P: ");
+          for (String key : grammar.getP().keySet()) {
+               System.out.println(buildString(key));
+          }
+     }
+
+     private static void printProductionsForANonTerminal(String nonTerminal) {
+          System.out.println("Productions for the non-terminal:");
+          System.out.println(buildString(nonTerminal));
+     }
+
+     private static String buildString(String key) {
+          List<List<String>> value = grammar.getP().get(key);
+          if (value != null) {
+               StringBuilder all = new StringBuilder();
+               for (List<String> list : value) {
+                    for (String s : list) {
+                         all.append(s);
+                    }
+                    all.append(" | ");
+               }
+               return "\t" + key + " -> " + all.substring(0, all.length() - 2);
+          } else return "NonTerminal not found\n";
+     }
 }
```

Comment:
outstanding code, 9.9/10, functional and very nicely written, have an easy exam session!